### PR TITLE
Verbosity and Logging

### PR DESCRIPTION
**Pull request by mxwt00**
Implemented the _--verbosity_ command flag and correct logging with log levels.
I had to alter the _--verbosity_ flag a bit. I could not get Clap's _parse(from_occurrences)_ to work correctly.
The log level is no longer increased with -v, -vv, ..., but instead by writing the log level as a String (for example _--verbosity debug_).

### CODE REVIEW

File: src/logging.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+use log::{info, LevelFilter};
+
+pub fn init_logging(verbose: &String) {
+    let level = match verbose.as_str() {
+        "error" => LevelFilter::Error,
+        "warn" => LevelFilter::Warn,
+        "info" => LevelFilter::Info,
+        "debug" => LevelFilter::Debug,
+        "trace" => LevelFilter::Trace,
+        _ => LevelFilter::Error //TODO throw an error
```
Comment (Danielprobst22) [inline src/logging.rs]: This todo has to be taken care of

File: src/logging.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+use log::{info, LevelFilter};
+
+pub fn init_logging(verbose: &String) {
+    let level = match verbose.as_str() {
```
Comment (Danielprobst22) [inline src/logging.rs]: We could rename the "verbose" variable and the flags for the commands to "verbosity"

**Comment by juliuskoskela** [comment]
I found this https://github.com/clap-rs/clap-verbosity-flag which implements standard behaviour for verobosity flag. What do you guys think?

**Comment by mxwt00** [comment]
> I found this https://github.com/clap-rs/clap-verbosity-flag which implements standard behavior for verbosity flag. What do you guys think?

Ah that should probably work. I personally like the variant with the log level as string more, but i can change it to use this variant if you prefer that.

**Comment by juliuskoskela** [comment]
In my opinion `-v***` is so standard that it would be the first expectation in a CLI without reading documentation. I think here it's justified to take the most standard approach from usability perspective.

File: src/import.rs [hunk 3]
```diff
@@ -30,26 +33,27 @@ pub struct Import {
...
+    ///Sets the CLI verbosity. Allowed values are 'error', 'warn', 'info', 'debug' and 'trace'.
+    #[clap(short, long, default_value = "error")]
+    pub verbose: String,
 }
 
 impl Import {
     //TODO old logic; has to be changed to match new commands
     pub fn import(&mut self) {
-        println!("Importing audio from SD card at {:?}", self.device);
+        init_logging(&self.verbose);
```
Comment (juliuskoskela) [inline src/import.rs]: Should we actually initiate logging in the main?

**Review by juliuskoskela — CHANGES_REQUESTED** [review]

**Comment by mxwt00** [comment]
Alright, then i will try out the clap_verbosity_flag

File: src/import.rs (hunk 3 above)
Comment (mxwt00) [reply src/import.rs]: thats not easily possible as far as i understand clap if the verbose flag is a flag of the subcommands

Comment (mxwt00) [reply src/import.rs]: but i think with the clap_verbosity_flag it should be possible, let me try that out

Comment (juliuskoskela) [reply src/import.rs]: Ahaa yes true should the verbosity flag actually be at toplevel?

File: src/main.rs [hunk 4]
```diff
@@ -20,14 +21,23 @@ pub enum Commands {
...
+
+    #[command(flatten)]
+    pub verbose: Verbosity,
 }
 
 
 fn main() {
-    let commands = Cli::parse();
+    let commands = Cli::parse().commands;
+    let verbose = Cli::parse().verbose;
```
Comment (Danielprobst22) [inline src/main.rs]: Can be simplified to
`let Cli { commands, verbose } = Cli::parse();`

File: src/main.rs [hunk 5]
```diff
@@ -20,14 +21,23 @@ pub enum Commands {
...
+    env_logger::builder().filter_level(verbose.log_level_filter()).init();
 
-    match commands.commands {
-        Commands::Import(mut import) => { import.import() }
+    match commands {
+        Commands::Import(mut import) => {
+            import.import()
+        }
         Commands::Clean(mut clean) => { clean.clean() }
     }
```
Comment (Danielprobst22) [inline src/main.rs]: For simple matches we don't need curly braces
```
    match commands {
        Commands::Import(mut import) => import.import(),
        Commands::Clean(mut clean) => clean.clean(),
    }
```

**Review by juliuskoskela — APPROVED** [review]

**Pull request merged**